src/harness/sas.py

```diff
@@ -304,9 +304,124 @@ def InjectCpiUser(self, request):
 
+def GetServer():
+  config_parser = ConfigParser.RawConfigParser()
+  config_parser.read(['sas.cfg'])
+  server_base_url = config_parser.get('HttpServer', 'BaseUrl')
+  server_port = int(config_parser.get('HttpServer', 'Port'))
+  return HttpServer({'baseUrl': server_base_url, 'port': server_port})
+
+class HttpServer(sas_interface.HttpServerInterface):
+  """Test Harness acting as a Http Server to receive Pull/GET and Push/POST
+   Requests from SAS Under Test
+  """
+  def __init__(self, server_details):
+    self._server_details = server_details
+    self._request_event = threading.Event()
+    self.parameters = {}
+    self.response = None
+
+  def _callbackHandler(self, response=None):
+    """Release wait initiated by _requestResponse when there is GET/POST Request from
+    SAS Under Test"""
+    self.response = response
+    self._request_event.set()
+
+  def _handleRequest(self, callbackHandler, getSetupParameters):
+    return lambda *args: HttpServerHandler(callbackHandler, getSetupParameters, *args)
+
+  def _getSetupParameters(self):
+    return self.parameters
+
+  def _requestResponse(self):
+    """Initiate wait till timeout to get the response from SAS Under Test and then
+    return the response as soon as _callbackHandler is called"""
+    logging.info("Waiting for the Response from SAS Under Test")
+    self._request_event.wait(20.0)
+    return self.response
+
+  def setupServer(self, parameters=None):
+    self.parameters = parameters
+    return self._requestResponse
+
+  def getBaseUrl(self):
+    return 'https://%s:%d' % (self._server_details['baseUrl'], self._server_details['port'])
+
+  def StartServer(self):
+    request_handler = self._handleRequest(self._callbackHandler, self._getSetupParameters)
+    self.server = HTTPServer((self._server_details['baseUrl'], self._server_details['port']),request_handler)
+    cert_file = 'server.cert'
+    key_file = 'server.key'
+    ca_cert = 'ca.cert'
+    ciphers = ['AES128-GCM-SHA256', 'AES256-GCM-SHA384', 'ECDHE-RSA-AES128-GCM-SHA256']
+    self.server.socket = ssl.wrap_socket(
+      self.server.socket,
+      certfile=cert_file,
+      keyfile=key_file,
+      ca_certs=ca_cert,
+      cert_reqs=ssl.CERT_REQUIRED,
+      ssl_version=ssl.PROTOCOL_TLSv1_2,
+      ciphers=':'.join(ciphers),
+      server_side=True
+      )
+    logging.info('Started Test Harness Server at %s' % self.getBaseUrl())
+    # Start Server in a separate thread
+    server_start = threading.Thread(target=self.server.serve_forever)
+    server_start.daemon = True
+    server_start.start()
+
+  def StopServer(self):
+    logging.info('Stopped Test Harness Server')
+    self.server.shutdown()
+
+class HttpServerHandler(BaseHTTPRequestHandler):
+  def __init__(self, callbackHandler, getSetupParameters, *args):
+    self._parameters = getSetupParameters()
+    self._callbackHandler = callbackHandler
+    BaseHTTPRequestHandler.__init__(self, *args)
+
+  def do_GET(self):
```

Comment:
In addition to the full dump message itself, we will also need to send back the files containing the CBSDs, PPAs, and ESC sensors. At first glance, this method doesn't appear to support that.
  
Edit: I see that you're setting up a separate instance for each piece of data that's included in the SAS TH. I'm not sure this is a sustainable long-term solution and it's a bit hard to read. I thought the solution we discussed at the F2F was to give the SAS TH a local directory from which it would serve files (if there's a local file whose name matches the requested path, it's served; otherwise HTTP 404). Then we can simply swap out files locally and leave the SAS TH running for the duration of the test. Is there a reason you chose to deviate from that?

Reply:
I rewritten the server in the new code. The server will be able to handle different response for CBSD, PPA, ESC or Full activity dump. If the path contains"cbsd_activity" it will return cbsd response. If the path contains"ppa_activity" it will return PPAS..